UART failure with CONFIG_UART_ASYNC_API and DMA

FRASTM opened the pull request:
when enabling the DMA of the stm32f7 soc,  especially on the tests/drivers/uart/uart_async_api, 
the cache DMA is not coherent.
--> Same approach than the https://github.com/zephyrproject-rtos/zephyr/pull/27911/
This PR is to help for fixing of  the https://github.com/zephyrproject-rtos/zephyr/issues/31711
This still give an error on the `test_chained_write`  test case

Signed-off-by: Francois Ramu francois.ramu@st.com

FRASTM committed:
tests: drivers: uart stm32 run uart_async api on nucleo_f767

This sets the dts of dma for using the uart 6 asynch api
The stm32f767 has a dma V1 with request 5 for Tx/Rx usart6
The Tx&Rx pins (PG14, PG9) of the usart6 are connected
on the nucleo_f767zi board to pass the test

Signed-off-by: Francois Ramu <francois.ramu@st.com>

dcpleung commented on drivers/serial/Kconfig.stm32:
```diff
@@ -20,3 +20,17 @@ config UART_STM32
+	default y
+	help
+	  Support unaligned DMA buffer. When DMA buffers are not aligned on
+	  cache line boundaries, or there size is not a multiple of the
```
s/there/their/

FRASTM replied:
@dcpleung I do not catch your point, sorry. This text is coming from the https://github.com/zephyrproject-rtos/zephyr/pull/27911/commits/70e111c1ae19e3aabfa1a95e906d40021542444b

dcpleung reviewed (commented):
There are couple commented out code block which needs to be removed.

FRASTM committed:
tests: drivers: uart stm32 run uart_async api on f746 nucleo

This sets the dts of dma for using the uart 6 asynch api.
The stm32f746 has a dma V1 with request 5 for Tx/Rx usart6
The Tx&Rx pins (PG14, PG9) of the usart6 are connected
on the nucleo_f746zg board to pass the test

Signed-off-by: Francois Ramu <francois.ramu@st.com>

dcpleung commented on drivers/serial/uart_stm32.c:
```diff
@@ -668,6 +668,14 @@ static inline void async_evt_rx_rdy(struct uart_stm32_data *data)
 		uart_stm32_invalidate_cache(
 			data->dma_rx.buffer,
 			data->dma_rx.buffer_length);
+/*
+		if (data->dma_rx.buffer != event.data.rx.buf) {
+			memcpy(event.data.rx.buf,
+				data->dma_rx.dma_aligned_buffer,
+				data->dma_rx_buffer_len);
+		}
+*/
```
Please remove commented out code blocks.

dcpleung commented on drivers/serial/uart_stm32.c:
```diff
@@ -1017,13 +1026,29 @@ static int uart_stm32_async_tx(const struct device *dev,
 	if (tx_data != NULL) {
+		uint32_t addr_diff = UART_STM32_ROUND_UP(tx_data)
+							- (uint32_t)tx_data;
+/*
+		if (addr_diff == 0 && buf_size >= __SCB_DCACHE_LINE_SIZE) {
+			buf_size = UART_STM32_ROUND_DOWN(buf_size);
+		} else {
+			if (addr_diff != 0) {
+				buf_size = MIN(buf_size, addr_diff);
+			}
+
+			buf_size = MIN(buf_size, __SCB_DCACHE_LINE_SIZE);
+
+			memcpy(data->dma_tx.dma_aligned_buffer,
+					data->dma_tx.buffer, buf_size);
+			tx_data = data->dma_tx.dma_aligned_buffer;
+		}
+*/
```
Remove commented out code block

dcpleung commented on drivers/serial/uart_stm32.c:
```diff
@@ -1077,6 +1102,28 @@ static int uart_stm32_async_rx_enable(const struct device *dev,
+	if (rx_buf != NULL) {
+		uint32_t addr_diff = UART_STM32_ROUND_UP(rx_buf)
+							- (uint32_t)rx_buf;
+/*
+		if (addr_diff == 0 && buf_size >= __SCB_DCACHE_LINE_SIZE) {
+			buf_size = UART_STM32_ROUND_DOWN(buf_size);
+		} else {
+			if (addr_diff != 0) {
+				buf_size = MIN(buf_size, addr_diff);
+			}
+
+			buf_size = MIN(buf_size, __SCB_DCACHE_LINE_SIZE);
+			rx_buf = data->dma_rx.dma_aligned_buffer;
+		}
+*/
```
Remove commented out code block

FRASTM committed:
soc: arm: stm32f7 cache config is configured by cortex_m

Remove the cache configuration from the stm32f7 soc,
as the core/aarch32/cortex_m inhibits it.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
initial issue is now closed
